t/t1092-sparse-checkout-compatibility.sh

```diff
@@ -1230,6 +1230,19 @@ test_expect_success 'sparse index is not expanded: update-index' '
 	ensure_not_expanded update-index --add --remove --again
 '
 
+test_expect_success 'sparse index is not expanded: read-tree' '
+	init_repos &&
+
+	ensure_not_expanded checkout -b test-branch update-folder1 &&
+	for MERGE_TREES in "base HEAD update-folder2" \
+		      "HEAD update-folder2" \
+		      "update-folder2"
+	do
```

Comment:
One case that might be interesting is a situation where we do a two- or three-way merge and two of the refs changed files such as `folder1/a` and `folder1/b`, respectively. This would make them look like they have an edit-edit conflict on the sparse directory entry, but if we expand the directory there would not be such a conflict.

Reply:
Ah, I completely missed this case (and you're right - it does cause a conflict). Since the conflict is only detected once you're already in `twoway_merge` or `threeway_merge` (in the middle of `traverse_trees`), you wouldn't be able to expand the index to resolve that conflict. 

I _think_ one way to handle this might be the way `oneway_diff` does (using `diff_tree_oid` to get the differences at a file level, then reconstructing the two- and three-way merges from series of one-way diffs). Since you wrote that addition to `oneway_diff`, I wanted to check - does that make sense? Or were you thinking something different for solving this?

Reply:
I _think_ that's the best way to attempt handling this. The tricky part is constructing the trees again afterwards, but maybe you can plug into the merge machinery to merge the trees to get a new tree? The ORT API might be a way to go here.

The other idea would be to report that the operation failed due to a sparse directory, then expand and try again. This isn't a great idea, but would be simpler to implement.

Reply:
I added a _tentative_ implementation for handling two-way merges (which, ultimately, are merging three trees: the index, input_tree_a, and input_tree_b). It passes the tests, but I'm still worried about all of the different edge cases of merges that could happen (e.g., `merge-ort` might successfully merge the contents of a file modified in two branches, but the existing `twoway_merge` might not). And, I haven't added anything yet for `threeway_merge` since the `n` number of base trees (very specific index conditions) don't currently fit nicely with `merge-ort`.

I do still want to avoid excessive index expansion, though - I _think_ one possibility is to refactor `unpack_trees` itself to, rather than using an element-by-element merge function, use `merge-ort` directly (when `unpack_trees_options.merge` is set). It would also require an extension to `merge-ort` to handle multiple base trees in a non-recursive merge. If it's not a terrible idea, I think it's doable - however, it's pretty substantial and probably warrants its own branch/patch series.

In any case, please look over what I have and let me know if you have any suggestions on direction. It's also potentially possible to wait on the two-way and three-way merge cases, since it looks like stash only ever calls `read-tree` with one tree argument.